### PR TITLE
[4] Empty State for System-> Discover

### PR DESCRIPTION
**Pull request by PhilETaylor**
### Summary of Changes

Empty state for when no extensions are discovered.

MOST OF THE TIME you will expect to see the Empty State. Its rare to see extensions needing discovering. This is the opposite of most components in Joomla admin. 

### Testing Instructions

Apply PR

Load System -> discover,

See Empty State 

click discover... or the button in the middle

Note new message. Nothing found. 

upload a plugin/extension manually into its right place using FTP (or copy and paste on your local machine) 

Load System -> discover, click discover... 

See you see the old screen where you can install the extension

### Actual result BEFORE applying this Pull Request

Screenshot 2021-04-28 at 23 46 50

### Expected result AFTER applying this Pull Request

Screenshot 2021-04-28 at 23 43 50

When you click either discover button you get a new message, or redirected to the list of discovered extensions to install (as normal) 

Screenshot 2021-04-28 at 23 43 56

### Documentation Changes Required

none

### CODE REVIEW

**Comment by PhilETaylor** [comment]
And yes I hate the title and sentences - please suggestion new ones if you dont like them too. I'll not be offended.

File: administrator/language/en-GB/com_installer.ini [hunk 1]
```diff
@@ -25,8 +25,12 @@ COM_INSTALLER_DISCOVER_TABLE_CAPTION="Table of Discovered Extensions"
 COM_INSTALLER_DOWNLOADKEY_EXTRA_QUERY_LABEL="Download Key"
 COM_INSTALLER_DOWNLOADKEY_MISSING_LABEL="The Download Key is missing"
 COM_INSTALLER_DOWNLOADKEY_MISSING_TIP="Updates for this extension will not work until you enter its Download Key. Click on the Update Site's title to enter the Download Key."
+COM_INSTALLER_EMPTYSTATE_BUTTON_ADD="Discover extensions to install"
+COM_INSTALLER_EMPTYSTATE_CONTENT="We have not found any extension on your site, that have not gone through the normal installation process."
+COM_INSTALLER_EMPTYSTATE_TITLE="You have no discovered extensions to install"
```
Comment (Quy) [inline administrator/language/en-GB/com_installer.ini]:
```suggestion
COM_INSTALLER_EMPTYSTATE_TITLE="You have no discovered extensions to install."
```

File: administrator/components/com_installer/src/View/Discover/HtmlView.php [hunk 2]
```diff
@@ -23,6 +23,14 @@
  */
 class HtmlView extends InstallerViewDefault
 {
+	/**
+	 * Is this view an Empty State
+	 *
+	 * @var  boolean
+	 * @since __DEPLOY_VERSION__
+	 */
+	public $isEmptyState = false;
```
Comment (Quy) [inline administrator/components/com_installer/src/View/Discover/HtmlView.php]:
```suggestion
	private $isEmptyState = false;
```

File: administrator/language/en-GB/com_installer.ini [hunk 3]
```diff
@@ -25,8 +25,12 @@ COM_INSTALLER_DISCOVER_TABLE_CAPTION="Table of Discovered Extensions"
 COM_INSTALLER_DOWNLOADKEY_EXTRA_QUERY_LABEL="Download Key"
 COM_INSTALLER_DOWNLOADKEY_MISSING_LABEL="The Download Key is missing"
 COM_INSTALLER_DOWNLOADKEY_MISSING_TIP="Updates for this extension will not work until you enter its Download Key. Click on the Update Site's title to enter the Download Key."
+COM_INSTALLER_EMPTYSTATE_BUTTON_ADD="Discover extensions to install"
+COM_INSTALLER_EMPTYSTATE_CONTENT="We have not found any extension on your site, that have not gone through the normal installation process."
```
Comment (brianteeman) [inline administrator/language/en-GB/com_installer.ini]:
```suggestion
COM_INSTALLER_EMPTYSTATE_CONTENT="We have not found any extensions on your site that have not gone through the normal installation process."
```

pluralise and remove comma

Not sure about the "We .." as that is not consistent with any other message on the site. Who is we?

**Review by brianteeman — CHANGES_REQUESTED** [review]

File: administrator/language/en-GB/com_installer.ini (hunk 3 above)
Comment (PhilETaylor) [reply administrator/language/en-GB/com_installer.ini]: [Like I said](https://github.com/joomla/joomla-cms/pull/33400#issuecomment-828831156) "And yes I hate the title and sentences too! - please suggestion new ones if you dont like them too. I'll not be offended."

File: administrator/components/com_installer/tmpl/discover/emptystate.php [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+use Joomla\CMS\Session\Session;
+
+$displayData = [
+	'textPrefix' => 'COM_INSTALLER',
+	'formURL'    => 'index.php?option=com_installer&task=discover.refresh',
+	'helpURL'    => 'https://docs.joomla.org/Help4.x:Extensions:_Discover',
+	'icon'       => 'icon-puzzle-piece install',
+];
+
+$user = Factory::getApplication()->getIdentity();
```
Comment (Quy) [inline administrator/components/com_installer/tmpl/discover/emptystate.php]: Delete?

File: administrator/components/com_installer/tmpl/discover/emptystate.php [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+$displayData['createURL'] = 'index.php?option=com_installer&task=discover.refresh&' . Session::getFormToken() . '=1';
```
Comment (Quy) [inline administrator/components/com_installer/tmpl/discover/emptystate.php]: Combine with the above array?

Comment (PhilETaylor) [reply administrator/components/com_installer/tmpl/discover/emptystate.php]: done

File: administrator/components/com_installer/tmpl/discover/emptystate.php (hunk 4 above)
Comment (PhilETaylor) [reply administrator/components/com_installer/tmpl/discover/emptystate.php]: done

File: administrator/language/en-GB/com_installer.ini [hunk 6]
```diff
@@ -25,8 +25,12 @@ COM_INSTALLER_DISCOVER_TABLE_CAPTION="Table of Discovered Extensions"
 COM_INSTALLER_DOWNLOADKEY_EXTRA_QUERY_LABEL="Download Key"
 COM_INSTALLER_DOWNLOADKEY_MISSING_LABEL="The Download Key is missing"
 COM_INSTALLER_DOWNLOADKEY_MISSING_TIP="Updates for this extension will not work until you enter its Download Key. Click on the Update Site's title to enter the Download Key."
+COM_INSTALLER_EMPTYSTATE_BUTTON_ADD="Discover extensions to install"
+COM_INSTALLER_EMPTYSTATE_CONTENT="We have not found any extensions on your site that have not gone through the normal installation process."
```
Comment (Quy) [inline administrator/language/en-GB/com_installer.ini]: The title states that there are no discovered extensions to install. There is no need to repeat it. Let's reuse the text from this string.

> COM_INSTALLER_MSG_DISCOVER_DESCRIPTION="Discover extensions that have not gone through the normal installation process."

Comment (PhilETaylor) [reply administrator/language/en-GB/com_installer.ini]: Done

File: administrator/components/com_installer/src/Model/DiscoverModel.php [hunk 7]
```diff
@@ -270,4 +270,30 @@ public function purge()
...
+			->clear('select')
+			->clear('values')
+			->clear('bounded')
+			->clear('limit')
+			->clear('order')
+			->clear('where')
+			->clear('where')
+			->select('count(*)');
+
+		$sql->where($this->_db->quoteName('state') . ' = ' . $this->_db->quote('-1'));
```
Comment (Quy) [inline administrator/components/com_installer/src/Model/DiscoverModel.php]:
```suggestion
		$sql->where($this->_db->quoteName('state') . ' = -1'));
```

File: administrator/components/com_installer/src/Model/DiscoverModel.php [hunk 8]
```diff
@@ -270,4 +270,30 @@ public function purge()
...
+			->clear('select')
+			->clear('values')
+			->clear('bounded')
+			->clear('limit')
+			->clear('order')
+			->clear('where')
+			->clear('where')
+			->select('count(*)');
+
+		$sql->where($this->_db->quoteName('state') . ' = -1'));
```
Comment (Quy) [inline administrator/components/com_installer/src/Model/DiscoverModel.php]:
```suggestion
		$sql->where($this->_db->quoteName('state') . ' = -1')));
```

Comment (PhilETaylor) [reply administrator/components/com_installer/src/Model/DiscoverModel.php]: nope - other way - needs less :)

File: administrator/components/com_installer/src/Model/DiscoverModel.php [hunk 9]
```diff
@@ -270,4 +270,30 @@ public function purge()
...
+			->clear('select')
+			->clear('values')
+			->clear('bounded')
+			->clear('limit')
+			->clear('order')
+			->clear('where')
+			->clear('where')
+			->select('count(*)');
+
+		$sql->where($this->_db->quoteName('state') . ' = -1')));
```
Comment (Quy) [inline administrator/components/com_installer/src/Model/DiscoverModel.php]:
```suggestion
		$sql->where($this->_db->quoteName('state') . ' = -1');
```

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 68479a232edd7d7e8638c143395249b21043bbee<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33400">issues.joomla.org/tracker/joomla-cms/33400</a>.</sub>

**Comment by drmenzelit** [comment]
I have tested this item :white_check_mark: successfully on 68479a232edd7d7e8638c143395249b21043bbee<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33400">issues.joomla.org/tracker/joomla-cms/33400</a>.</sub>

**Comment by PhilETaylor** [comment]
This has conflicts (because com_installer has multiple empty states, and thus has language string conflicts) and cannot be merged just yet - I'll fix it once enough coffee has been injected into my veins.

**Comment by PhilETaylor** [comment]
Conflicts resolved, merged in from 4.0-dev and ready for testing again. Thanks.

**Comment by PhilETaylor** [comment]
Hey @wilsonge If you are thinking of cutting another Joomla 4 beta very shortly can we quickly get this one and the other one I'll ping you merged, these are the last 2 Empty States and without them the feature is not yet fully implemented :-) Thanks.

**Comment by alikon** [comment]
missed from my radar  ... due to the  storm :smile:

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on 2b9078b2bbfa49cf7bfe4a698103b274028a7dd0<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33400">issues.joomla.org/tracker/joomla-cms/33400</a>.</sub>

**Comment by alikon** [comment]
fields and field groups are in your radar @PhilETaylor  ??

**Comment by PhilETaylor** [comment]
> fields and field groups are in your radar @PhilETaylor ??

They are hard, due to the additional context and dropdown needed to make it make sense as an empty state

**Comment by PhilETaylor** [comment]
>  ... due to the storm

We had the french try to invade our island yesterday.. such fun...

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 2b9078b2bbfa49cf7bfe4a698103b274028a7dd0<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33400">issues.joomla.org/tracker/joomla-cms/33400</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33400">issues.joomla.org/tracker/joomla-cms/33400</a>.</sub>

**Comment by Quy** [comment]
Thanks you!!!

**Pull request merged**